### PR TITLE
acond: Add signal/kill interface

### PR DESCRIPTION
**Pull request by xiangquanliu**

### CODE REVIEW

File: acond/src/utils.rs [hunk 1]
```diff
@@ -41,9 +41,9 @@ pub const ERR_RPC_REJECT_BLOB: &str = "Reject blob as no existing manifests requ
 pub const ERR_RPC_INVALID_IMAGE_ID: &str = "Invalid image id.";
 pub const ERR_RPC_INVALID_CONTAINER_ID: &str = "Invalid container id.";
 pub const ERR_RPC_CONTAINER_EXITED: &str = "Container has exited.";
-pub const ERR_RPC_CONTAINER_NOT_ALLOW_KILL: &str = "Container is not allowed to be killed.";
-pub const ERR_RPC_CONTAINER_KILL_TIMEOUT: &str = "Container can't be killed in {} seconds.";
+pub const ERR_RPC_CONTAINER_RESTART_TIMEOUT: &str = "Container can't be restarted in {} seconds.";
 pub const ERR_RPC_CONTAINER_NOT_ALLOW_RESTART: &str = "Container is not allowed to be restarted.";
+pub const ERR_RPC_CONTAINER_NOT_ALLOW_KILL: &str = "Container is not allowed to be killed.";
```
Comment (binxing) [inline acond/src/utils.rs:46]: `"Signal not allowed"`?

Comment (xiangquanliu) [reply acond/src/utils.rs:46]: So as discussed, all these error messages should be refactored as the same style - like "Timeout restarting container" and "Signal not allowed"?

File: acond/src/utils.rs [hunk 2]
```diff
@@ -30,31 +30,31 @@ use std::{
...
+pub const ERR_RPC_INVALID_MALIAS_FORMAT: &str = "Invalid manifest alias format";
 #[cfg(not(feature = "interactive"))]
-pub const ERR_RPC_NOT_SUPPORT_IA_MODE: &str = "Not support interactive mode.";
-pub const ERR_RPC_FAIL_FORK: &str = "Fails to execute fork.";
-pub const ERR_RPC_BUFFER_EXCEED: &str = "Stdin buffer size exceeds {} bytes.";
-pub const ERR_RPC_INVALID_COMMAND: &str = "Command '{}' doesn't start with a capital letter.";
-pub const ERR_RPC_SYSTEM_ERROR: &str = "System error, errno: {}.";
-pub const ERR_IPC_INVALID_REQ_FORMAT: &str = "Invalid request struct format.";
-pub const ERR_IPC_NOT_SUPPORT_REQ: &str = "Not support request.";
+pub const ERR_RPC_NOT_SUPPORT_IA_MODE: &str = "Interactive mode not supported";
```
Comment (binxing) [inline acond/src/utils.rs:51]: How about `ERR_RPC_NOT_SUPPORTED = "Not supported"`?

File: acond/src/utils.rs [hunk 3]
```diff
@@ -30,31 +30,31 @@ use std::{
...
+pub const ERR_RPC_FAIL_FORK: &str = "Fail to execute fork";
+pub const ERR_RPC_BUFFER_EXCEED: &str = "Stdin buffer size exceeds capture size";
+pub const ERR_RPC_INVALID_COMMAND: &str = "Command not start with a capital letter";
```
Comment (binxing) [inline acond/src/utils.rs:54]: We distinguish public vs private entry points using the first letter. So here how about `PRIVATE_ENTRYPOINT = "Private entry point"`?

File: acond/src/utils.rs [hunk 4]
```diff
@@ -30,31 +30,31 @@ use std::{
...
 #[cfg(not(feature = "interactive"))]
-pub const ERR_RPC_NOT_SUPPORT_IA_MODE: &str = "Not support interactive mode.";
-pub const ERR_RPC_FAIL_FORK: &str = "Fails to execute fork.";
-pub const ERR_RPC_BUFFER_EXCEED: &str = "Stdin buffer size exceeds {} bytes.";
-pub const ERR_RPC_INVALID_COMMAND: &str = "Command '{}' doesn't start with a capital letter.";
-pub const ERR_RPC_SYSTEM_ERROR: &str = "System error, errno: {}.";
-pub const ERR_IPC_INVALID_REQ_FORMAT: &str = "Invalid request struct format.";
-pub const ERR_IPC_NOT_SUPPORT_REQ: &str = "Not support request.";
+pub const ERR_RPC_NOT_SUPPORT_IA_MODE: &str = "Interactive mode not supported";
+pub const ERR_RPC_FAIL_FORK: &str = "Fail to execute fork";
```
Comment (binxing) [inline acond/src/utils.rs:52]: Isn't this one of `SYSTEM_ERROR`?

File: acond/src/utils.rs [hunk 5]
```diff
@@ -30,31 +30,31 @@ use std::{
...
+pub const ERR_RPC_BUFFER_EXCEED: &str = "Stdin buffer size exceeds capture size";
+pub const ERR_RPC_INVALID_COMMAND: &str = "Command not start with a capital letter";
+pub const ERR_RPC_SYSTEM_ERROR: &str = "System error, errno: {}";
+pub const ERR_IPC_INVALID_REQ_FORMAT: &str = "Invalid request struct format";
```
Comment (binxing) [inline acond/src/utils.rs:56]: `INVALID_REQUEST`?

File: acond/src/utils.rs [hunk 6]
```diff
@@ -30,31 +30,31 @@ use std::{
...
+pub const ERR_IPC_NOT_SUPPORT_REQ: &str = "Request not supported";
```
Comment (binxing) [inline acond/src/utils.rs:57]: How about `ERR_IPC_NOT_SUPPORTED`?

**Review by binxing — CHANGES_REQUESTED** [review]

File: acond/src/utils.rs [hunk 7]
```diff
@@ -30,31 +30,31 @@ use std::{
...
-pub const ERR_RPC_INVALID_IMAGE_ID: &str = "Invalid image id.";
-pub const ERR_RPC_INVALID_CONTAINER_ID: &str = "Invalid container id.";
-pub const ERR_RPC_CONTAINER_EXITED: &str = "Container has exited.";
-pub const ERR_RPC_CONTAINER_NOT_ALLOW_KILL: &str = "Container is not allowed to be killed.";
-pub const ERR_RPC_CONTAINER_KILL_TIMEOUT: &str = "Container can't be killed in {} seconds.";
-pub const ERR_RPC_CONTAINER_NOT_ALLOW_RESTART: &str = "Container is not allowed to be restarted.";
-pub const ERR_RPC_NO_IMAGES: &str = "No images in current TD.";
-pub const ERR_RPC_INVALID_LPOLICY_FORMAT: &str = "Invalid launch policy format.";
-pub const ERR_RPC_INVALID_MALIAS_FORMAT: &str = "Invalid manifest alias format.";
+pub const ERR_CFG_INVALID_VSOCK_PORT: &str = "Invalid kernel parameter: vsock port";
```
[the remark this answers is not in the document]
Comment (xiangquanliu) [reply acond/src/utils.rs:33]: Agree to use "Invalid kernel cmdline parameter - acond.vsock_port"

File: acond/src/rpc.rs [hunk 8]
```diff
@@ -322,6 +334,50 @@ impl AconService for TDAconService {
...
+
+        if !container.is_running() {
+            return Err(Status::unknown(utils::ERR_RPC_CONTAINER_EXITED));
+        }
+
+        let image = pod
+            .get_image(&container.image_id)
+            .ok_or_else(|| Status::invalid_argument(utils::ERR_RPC_INVALID_IMAGE_ID))?;
+
+        if let None = image.manifest.signals.iter().find(|&&s| s == signal_num) {
```
[the remark this answers is not in the document]
Comment (xiangquanliu) [reply acond/src/rpc.rs:353]: From the arch docs, it should be an exact match.

File: acond/src/utils.rs [hunk 9]
```diff
@@ -41,9 +41,9 @@ pub const ERR_RPC_REJECT_BLOB: &str = "Reject blob as no existing manifests requ
 pub const ERR_RPC_INVALID_IMAGE_ID: &str = "Invalid image id.";
 pub const ERR_RPC_INVALID_CONTAINER_ID: &str = "Invalid container id.";
 pub const ERR_RPC_CONTAINER_EXITED: &str = "Container has exited.";
-pub const ERR_RPC_CONTAINER_NOT_ALLOW_KILL: &str = "Container is not allowed to be killed.";
-pub const ERR_RPC_CONTAINER_KILL_TIMEOUT: &str = "Container can't be killed in {} seconds.";
+pub const ERR_RPC_CONTAINER_RESTART_TIMEOUT: &str = "Container can't be restarted in {} seconds.";
```
[the remark this answers is not in the document]
Comment (xiangquanliu) [reply acond/src/utils.rs]: Agree.

File: acond/src/utils.rs (hunk 1 above)
Comment (xiangquanliu) [reply acond/src/utils.rs]: Agree

File: acond/src/utils.rs [hunk 10]
```diff
@@ -30,31 +30,31 @@ use std::{
...
-pub const ERR_RPC_CONTAINER_EXITED: &str = "Container has exited.";
-pub const ERR_RPC_CONTAINER_NOT_ALLOW_KILL: &str = "Container is not allowed to be killed.";
-pub const ERR_RPC_CONTAINER_KILL_TIMEOUT: &str = "Container can't be killed in {} seconds.";
-pub const ERR_RPC_CONTAINER_NOT_ALLOW_RESTART: &str = "Container is not allowed to be restarted.";
-pub const ERR_RPC_NO_IMAGES: &str = "No images in current TD.";
-pub const ERR_RPC_INVALID_LPOLICY_FORMAT: &str = "Invalid launch policy format.";
-pub const ERR_RPC_INVALID_MALIAS_FORMAT: &str = "Invalid manifest alias format.";
+pub const ERR_CFG_INVALID_VSOCK_PORT: &str = "Invalid kernel parameter: vsock port";
+pub const ERR_CFG_INVALID_TCPIP_PORT: &str = "Invalid kernel parameter: TCP/IP port";
+pub const ERR_CFG_INVALID_TIMEOUT: &str = "Invalid kernel parameter: timeout";
```
[the remark this answers is not in the document]
Comment (xiangquanliu) [reply acond/src/utils.rs:35]: Agree.

File: acond/src/utils.rs [hunk 11]
```diff
@@ -30,31 +30,31 @@ use std::{
...
+pub const ERR_RPC_REJECT_MANIFEST: &str = "Manifest rejected according to the launch policy";
+pub const ERR_RPC_INVALID_SIGNATURE: &str = "Invalid digital signature";
+pub const ERR_RPC_INVALID_CERTIFICATE: &str = "Invalid certificate";
+pub const ERR_RPC_INVALID_HASH_ALGORITHM: &str = "Invalid hash algorithm";
+pub const ERR_RPC_REJECT_BLOB: &str = "Blob rejected as no manifests require the blob";
```
[the remark this answers is not in the document]
Comment (xiangquanliu) [reply acond/src/utils.rs:40]: Agree.

File: acond/src/utils.rs (hunk 2 above)
Comment (xiangquanliu) [reply acond/src/utils.rs:51]: Not sure it is suitable to only use "Not supported". The scenario is in line 251 of rpc.rs. Should it be changed to "Invalidate parameter" or something?

File: acond/src/utils.rs (hunk 3 above)
Comment (xiangquanliu) [reply acond/src/utils.rs:54]: Agree.

File: acond/src/utils.rs (hunk 4 above)
Comment (xiangquanliu) [reply acond/src/utils.rs:52]: Agree. Will use system error instead.

File: acond/src/utils.rs (hunk 5 above)
Comment (xiangquanliu) [reply acond/src/utils.rs:56]: Agree.

File: acond/src/utils.rs (hunk 6 above)
Comment (xiangquanliu) [reply acond/src/utils.rs:57]: Agree.

**Review by binxing — APPROVED** [review]

**Comment by binxing** [comment]
No further feedbacks from team members, so merge PR.

**Pull request merged**